Tbrands 27 algolia assortment content source

JackHowa opened the pull request:
## Description

Add algolia content source

## Jira Ticket

- [TMEDIA-27](https://arcpublishing.atlassian.net/browse/TMEDIA-27)

## Acceptance Criteria

A new block in the arc-themes-blocks repo - algolia-assortment-content-source

A Fusion Content Source that can query the Algolia Search API with the following parameters

Index

Context - Algolia Param - ruleContexts

Filters - Algolia Param - filters

Query - Algolia Param - query

Size - Algolia Param - hitsPerPage - default = 20

Page - Algolia Param - page - default = 0

The content source relies on ApplicationId and APIKey to set up the Algolia search methods

Content Source has TTL set to the Fusion minimum of 2 minutes

## Test Steps

1. Go to feature pack repo 
2. Checkout this branch `git checkout TBRANDS-27-algolia-assortment-content-source}`
3. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/algolia-assortment-content-source-block`
4. Go to blocks repo
5. Checkout this branch `git checkout TBRANDS-27-algolia-assortment-content-source}`
6. Go to feature pack repo.  https://github.com/WPMedia/arc-themes-feature-pack/pull/218
7. Ensure your .env has plaintext values for `algoliaSearchKey` and `algoliaAppId`. See https://github.com/WPMedia/arc-themes-feature-pack/pull/209#issue-1154629794 for getting these values if you don't have them already
8. Make sure you've added `@local` to the block name in the blocks.json
9. Run `npx fusion start -f -l @wpmedia/algolia-assortment-content-source-block`
10. Go to content debugger on local for algolia assortment http://localhost/pagebuilder/tools/debugger?_website=arc-commerce&type=content-debugger&source=algolia-assortment
11. Run a query like and notice the first image url. Notice there are 20 items returned. This means there are 20 items by default and it starts on page 0. `searchdemoproduct_commercedev_` for index, `shoes` for query, `featured_shoes` for rule contexts
![Screen Shot 2022-04-12 at 10 57 59](https://user-images.githubusercontent.com/5950956/163004309-79c4ae73-119b-4491-8c28-34a4d736520a.png)
12. Now set the same query to 21 count. You'll see 21 items
13. Now set the page to 1. You will see a different image url than by default before
![Screen Shot 2022-04-12 at 10 59 51](https://user-images.githubusercontent.com/5950956/163004842-d8dafcb6-bf8d-4327-8ebf-bd4ab70de549.png)
14. Now add a filter to notice the effect of adding a specific sku. (In our example, there's 3 items with the same sku.) The filter text should be: `("sku": "SHOE34YXGQYWM45Y")`
![Screen Shot 2022-04-12 at 11 01 49](https://user-images.githubusercontent.com/5950956/163005190-87a2fc63-fb36-4ead-9486-6458e7d9e382.png)

## Effect Of Changes

### Before

- No algolia assortment content source

### After

- have an algolia assortment content source! 

## Dependencies or Side Effects

- Need the environment files. I've added them in the feature pack pr for the dev commerce environment with the key encrypted
- Merge and deploy https://github.com/WPMedia/arc-themes-feature-pack/pull/218

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.


### Notes 
- Go to https://github.com/WPMedia/arc-themes-feature-pack/pull/209#issue-1154629794 for getting the algoliaSearchKey set in `.env`. The `algoliaAppId` should also be set in the `.env` 
- Add     "@wpmedia/algolia-assortment-content-source-block@local" to blocks.json. run `npx fusion start -f -l  @wpmedia/algolia-assortment-content-source-block` to get this working locally
- I've added additional comments after feedback from @marissa-halpert 

Working example with options set from ticket 

index
searchdemoproduct_commercedev_

<img width="648" alt="Screen Shot 2022-04-11 at 14 05 05" src="https://user-images.githubusercontent.com/5950956/162813152-bda4940f-07b4-4ffe-8da9-95af01bdd372.png">



### Questions
- I don't seem to be able to use filters. I've tried various configurations https://www.algolia.com/doc/api-reference/api-parameters/filters/ , including is_fk_advantage_product:false and  "is_fk_advantage_product":"false". Is there something that I'm missing with the format or the data? Looking for help on this end as I did not see it in Nick's pr 
- For the algolia app id, I set this in the environment file. I'm assuming this is ok based on the ticket ac https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TBRANDS-27 Is this ok? 


### Todos

- [x] Update readme with basic info and params
- [x] Create feature pack pr with the environment values, including encrypted search key and plaintext app id
- [x] Add tests with mocking algolia api

JackHowa committed:
ci: Update generator for single content source

JackHowa committed:
ci: Initially generate the content source algolia assortment

JackHowa committed:
feat: Install algolia search lib. Add basic environment secret mocking.

JackHowa committed:
feat: Take in filters and other params to query algolia api

JackHowa committed:
test: Add basic mock and test to algolia assortment block

matthewroach commented:
@JackHowa - This is working for me

<img width="792" alt="Microsoft Edge-2022-04-11-22-05 28" src="https://user-images.githubusercontent.com/868127/162832120-cbf09c4e-41cc-43e7-9947-1dc79afd8950.png">

JackHowa commented:
> @JackHowa - This is working for me
> 
> <img alt="Microsoft Edge-2022-04-11-22-05 28" width="792" src="https://user-images.githubusercontent.com/868127/162832120-cbf09c4e-41cc-43e7-9947-1dc79afd8950.png">

huh cool. I'll work on this and get it ready in a bit

JackHowa committed:
test: Add mock and error-handling tests for algolia lib

JackHowa committed:
feat: Add hits per page and page parameters

JackHowa committed:
feat: Add fallback values for null values with page and results per page with new requirements

JackHowa committed:
docs: Add basic info on algolia content source block

matthewroach commented:
@JackHowa - Can you update the package files to ensure the actions are not failing please

JackHowa committed:
ci: Ensure package lock in sync

JackHowa commented on _templates/content-source/new/default-test.ejs.t:
```diff
@@ -4,7 +4,7 @@ to: blocks/<%= h.inflection.dasherize(block_name) %>-content-source-block/source
 import contentSource from './<%= h.inflection.dasherize(block_name) %>';
 
-describe('Test <%= h.changeCase.title(content_source_name) %> content source', () => {
+describe('Test <%= h.changeCase.title(block_name) %> content source', () => {
```
noticed issue with this naming and fixed

JackHowa committed:
ci: Ensure algolia is a dev dep like our other test deps

JackHowa committed:
ci: Reinstall deps

matthewroach commented on blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:
```diff
@@ -0,0 +1,49 @@
+};
+
+const fetch = async (contentSourceParameters) => {
+	const { filters, index, query, ruleContexts } = contentSourceParameters;
+	let { hitsPerPage, page } = contentSourceParameters;
+
+	// parameters in content source are defined as null by default
+	if (hitsPerPage === null) {
+		hitsPerPage = 20;
+	}
+	if (page === null) {
+		page = 0;
+	}
```
Any reason not to set defaults like so ?

```suggestion
	const {
		filters,
		index,
		query,
		ruleContexts,
		hitsPerPage = 20,
		page = 0,
	} = contentSourceParameters;
```

JackHowa replied:
yes. in the comment and test, I said the values were coming in as `null`. So we can't use default parameters @matthewroach

matthewroach commented on blocks/algolia-assortment-content-source-block/README.md:
```diff
@@ -0,0 +1,26 @@
+
+The Algolia assortment content source block allows you to display products from an Algolia index.
+
+## ANS Schema
```
Minor - but algolia is not using ANS Schema

JackHowa replied:
ah yeah. one sec

JackHowa replied:
updated @matthewroach

JackHowa committed:
docs: Update algolia info per comment https://github.com/WPMedia/arc-themes-blocks/pull/1312#discussion_r848846117

brentmiller1973 commented on blocks/algolia-assortment-content-source-block/README.md:
````diff
@@ -0,0 +1,26 @@
+## Configurable Params
+
+```
+    filters: "text",
````
Wondering if we could expand on the params a bit here?  More description and examples possibly?

JackHowa replied:
updated with some description and links to algolia's documentation. I don't want to recreate their docs too much

JackHowa replied:
Updated! thanks for the feedback @visionbegin. I added some links to algolia's documentation related. I don't want to recreate their docs too much (they have really great docs with examples).

brentmiller1973 replied:
That works for me!

brentmiller1973 replied:
Issue is resolved

ecarlisle commented on blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:
```diff
@@ -0,0 +1,49 @@
+	params,
+	transform: (data) => data,
+	// Time to live cache set to the Fusion minimum of 2 minutes (120 seconds)
+	ttl: 120,
```
@JackHowa : FYI as a topic to revisit but not a blocker. The Commerce team has mentioned being amenable to a 5-minute TTL for assortment data. Delivery may appreciate that as well.

JackHowa replied:
gotcha. I was going off of the ticket requirements. if that's not the goal, then I can change this @ecarlisle

matthewroach replied:
@ecarlisle Where was this mentioned? As we refined the ticket with the 2 minute ttl - I want to make sure we are all on the same page with decisions being made.
We can move forward with the 2 minute ttl for now, and create a ticket to update it to something else if needed

ecarlisle replied:
@matthewroach : Long ago during the initial consultations Algolia gave us. I agree with starting with 2 minutes and making a ticket for any future changes.

ecarlisle commented on blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:
```diff
@@ -0,0 +1,49 @@
+			filters,
+			hitsPerPage,
+			page,
+		});
```
Also for another day, but when the Commerce schema is final we can use this as a filter to minimize payload:
https://www.algolia.com/doc/api-reference/api-parameters/attributesToRetrieve/

JackHowa replied:
Yep good point. Is this ok to resolve? @ecarlisle

ecarlisle reviewed (approved):
Just added some notes for future but I think we're in a great place!

JackHowa committed:
docs: Add more links to syntax with algolia

Per https://github.com/WPMedia/arc-themes-blocks/pull/1312#discussion_r848867598
